pulsar-client-kafka-compat/src/main/java/org/apache/kafka/clients/consumer/PulsarKafkaConsumer.java

```diff
@@ -0,0 +1,482 @@
+            this.keyDeserializer = config.getConfiguredInstance(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+                    Deserializer.class);
+            this.keyDeserializer.configure(config.originals(), true);
+        } else {
+            this.keyDeserializer = keyDeserializer;
+            config.ignore(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG);
+        }
+
+        if (valueDeserializer == null) {
+            this.valueDeserializer = config.getConfiguredInstance(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+                    Deserializer.class);
+            this.valueDeserializer.configure(config.originals(), true);
+        } else {
+            this.valueDeserializer = valueDeserializer;
+            config.ignore(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG);
+        }
+
+        groupId = config.getString(ConsumerConfig.GROUP_ID_CONFIG);
+        isAutoCommit = config.getBoolean(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG);
+
+        String serviceUrl = config.getList(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG).get(0);
+
+        Properties properties = new Properties();
+        config.originals().forEach((k, v) -> properties.put(k, v));
+        ClientConfiguration clientConf = PulsarKafkaConfig.getClientConfiguration(properties);
+        try {
+            client = PulsarClient.create(serviceUrl, clientConf);
+        } catch (PulsarClientException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void received(org.apache.pulsar.client.api.Consumer consumer, Message msg) {
+        // Block listener thread if the application is slowing down
+        try {
+            receivedMessages.put(new QueueItem(consumer, msg));
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public Set<TopicPartition> assignment() {
+        throw new UnsupportedOperationException("Cannot access the partitions assignements");
+    }
+
+    /**
+     * Get the current subscription. Will return the same topics used in the most recent call to
+     * {@link #subscribe(Collection, ConsumerRebalanceListener)}, or an empty set if no such call has been made.
+     *
+     * @return The set of topics currently subscribed to
+     */
+    @Override
+    public Set<String> subscription() {
+        return consumers.keySet();
+    }
+
+    @Override
+    public void subscribe(Collection<String> topics) {
+        topics.forEach(topic -> {
+            try {
+                // Create individual subscription on each partition, that way we can keep using the
+                // acknowledgeCumulative()
+                PartitionedTopicMetadata partitionMetadata = ((PulsarClientImpl) client)
+                        .getPartitionedTopicMetadata(topic).get();
+
+                ConsumerConfiguration conf = new ConsumerConfiguration();
+                conf.setSubscriptionType(SubscriptionType.Failover);
```

Comment:
do the `failover` subscription have the listener when it becomes standby? so in that way, we can use that listener as the consumer rebalance listener.

Reply:
Yes, the client library has no knowledge whether it's active or standby, it will just stop getting messages being pushed. The listener will be there in any case.